gensim/models/keyedvectors.py

```diff
@@ -308,6 +308,7 @@ def add(self, entities, weights, replace=False):
             self.index2entity.append(entity)
 
         # add vectors for new entities
+        self.vectors = self.vectors.astype(weights.dtype)  # cast existing vectors to 'weights' type
```

Comment:
If I understand `astype()` correctly, this will always create a new copy of the underlying array (possibly quite wasteful for a large `vectors` or one that's already the proper type). Also, for consistency, shouldn't the type always be `float32`, rather than matching `weights`? So, wouldn't it be better to: (1) ensure even the empty initialization of 'vectors' just specifies the right `float32` type; & (2) always cast the passed-in `weights` (if necessary) to the proper type?

Reply:
> shouldn't the type always be float32, rather than matching weights? 

good question, to be honest, I'm not sure.

> (1) ensure even the empty initialization of 'vectors' in KeyedVectors.__init__() just specifies the right float32 type from the start;

I like that idea, maybe I'll add `dtype` argument to init function with the default `float32`, wdyt?

> (2) always cast the usually-smaller passed-in weights (as necessary) to the proper type

if we implement `(1)` - I agree

Reply:
I added datatype as parameter (with f32 default) - e031fc4

Reply:
Defaulting to `float32` is definitely good; I think there's a chance that letting it ever be anything else would cause problems, perhaps SEGFAULT/crashing problems, in any Cython code that might wind up operating on such not-as-expected vectors.

Reply:
Agree, in that case, I'll remove parametrization and just "pin" f32 everywhere (to avoid potential issues with Cython)

Reply:
Fixed 7c7ce19